### ISSUE TITLE
False positive mismatch types

### ISSUE THREAD

**Issue by mili-l** [intellij-rust/intellij-rust#8464]
## Environment

* **IntelliJ Rust plugin version:** 0.4.165.4391-213-nightly
* **Rust toolchain version:** 1.59.0-nightly (0b42deacc 2021-12-09) x86_64-apple-darwin
* **IDE name and version:** IntelliJ IDEA 2021.3.1 Ultimate Edition (IU-213.6461.79)
* **Operating system:** macOS 11.5.1
* **Macro expansion engine:** new
* **Name resolution engine:** new
* **Additional experimental features:** org.rust.cargo.features.settings.gutter, org.rust.cargo.evaluate.build.scripts, org.rust.macros.proc

## Problem description
False positive mismatch types

## Steps to reproduce
```rust
let y = &[0; 10];
    let p: *const [i32] = y;
```
<img width="829" alt="Screenshot 2022-01-25 at 13 09 00" src="https://user-images.githubusercontent.com/1213122/150974426-b749dc5d-0498-4550-91e9-1a9284f53b15.png">

**Comment by ccleve** [comment]
Related problem:

`let data: Box<[u8]> = Box::new([0u8; 256]);`

shows red squiggle under `[0u8; 256]` with the error 

```
mismatched types [E0308] 
expected `[u8]`, found `[u8; 256]` 
```

The code compiles and runs fine, though.

The suggested fix is "Convert to [u8] using *". When I click that, it gives me:

`let data: Box<[u8]> = Box::new(*[0u8; 256]);`

which gets past CLion, but the compiler says this:

```
let data: Box<[u8]> = Box::new(*[0u8; 256]);
                               ^^^^^^^^^^^ doesn't have a size known at compile-time
```

CLion, Rust plugin,Rust toolchain, and Mac OS are all current.

**Comment by otwieracz** [comment]
Another example, expecting _unknown_:
```
mismatched types [E0308] expected `&mut <unknown>`, found `Value` 
```

<img width="771" alt="Screenshot 2022-02-23 at 20 44 50" src="https://user-images.githubusercontent.com/108072/155396116-1ff8c6d5-61a8-4f4d-850e-0dcf906d8398.png">

**Comment by Undin** [comment]
@otwieracz could you file a separate issue with some reproducible example, please? Type inference issues have a lot of different reasons. And it looks like your issue is not the same as the original one
